token/program-2022/src/extension/mod.rs

```diff
@@ -1951,82 +2137,379 @@ mod test {
+        /// overallocated to mimic the runtime
+        fn new(account_data: &[u8]) -> Self {
+            let mut data = vec![];
+            data.extend_from_slice(&(account_data.len() as u64).to_le_bytes());
+            data.extend_from_slice(account_data);
+            data.extend_from_slice(&[0; MAX_PERMITTED_DATA_INCREASE]);
+            Self {
+                data,
+                lamports: 10,
+                owner: Pubkey::new_unique(),
+            }
+        }
+
+        /// Data lops off the first 8 bytes, since those store the size of the
+        /// account for the Solana runtime
+        fn data(&self) -> &[u8] {
+            let start = size_of::<u64>();
+            let len = self.len();
+            &self.data[start..start + len]
+        }
+
+        /// Gets the runtime length of the account data
+        fn len(&self) -> usize {
+            self.data
+                .get(..size_of::<u64>())
+                .and_then(|slice| slice.try_into().ok())
+                .map(u64::from_le_bytes)
+                .unwrap() as usize
+        }
+    }
+    impl GetAccount for SolanaAccountData {
+        fn get(&mut self) -> (&mut u64, &mut [u8], &Pubkey, bool, Epoch) {
+            // need to pull out the data here to avoid a double-mutable borrow
+            let start = size_of::<u64>();
+            let len = self.len();
+            (
+                &mut self.lamports,
+                &mut self.data[start..start + len],
+                &self.owner,
+                false,
+                Epoch::default(),
+            )
+        }
+    }
+
+    #[test]
+    fn alloc_new_tlv_in_account_info_from_base_size() {
+        let variable_len = VariableLenMintTest { data: vec![20, 99] };
+        let value_len = variable_len.get_packed_len().unwrap();
+        let base_account_size = Mint::LEN;
+        let mut buffer = vec![0; base_account_size];
+        let mut state = StateWithExtensionsMut::<Mint>::unpack_uninitialized(&mut buffer).unwrap();
+        state.base = TEST_MINT;
+        state.pack_base();
+
+        let mut data = SolanaAccountData::new(&buffer);
+        let key = Pubkey::new_unique();
+        let account_info = (&key, &mut data).into_account_info();
+
+        alloc_and_serialize::<Mint, _>(&account_info, &variable_len).unwrap();
+        let new_account_len = BASE_ACCOUNT_AND_TYPE_LENGTH + add_type_and_length_to_len(value_len);
+        assert_eq!(data.len(), new_account_len);
+        let state = StateWithExtensions::<Mint>::unpack(data.data()).unwrap();
+        assert_eq!(
+            state
+                .get_variable_len_extension::<VariableLenMintTest>()
+                .unwrap(),
+            variable_len
+        );
+
+        // alloc again fails
+        let account_info = (&key, &mut data).into_account_info();
+        assert_eq!(
+            alloc_and_serialize::<Mint, _>(&account_info, &variable_len).unwrap_err(),
+            TokenError::ExtensionAlreadyInitialized.into()
+        );
```

Comment:
I guess the question above (about unifying methods) is whether we need a failure on a second alloc.

Reply:
As long as we're clear on the behavior in the comment, then I'm happy to combine both! Since it's pretty low-level, I think it's perfectly legit to add a "thar be dragons here, be sure to check if you're planning reallocating" comment